airsonic-main/src/test/java/org/airsonic/player/service/search/SearchServiceSpecialGenreTestCase.java

```diff
@@ -0,0 +1,312 @@
+import org.junit.rules.TemporaryFolder;
+import org.junit.runner.Description;
+import org.junit.runners.model.Statement;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.core.io.ResourceLoader;
+import org.springframework.test.annotation.DirtiesContext;
+import org.springframework.test.context.ContextConfiguration;
+import org.springframework.test.context.junit4.rules.SpringClassRule;
+import org.springframework.test.context.junit4.rules.SpringMethodRule;
+import static org.springframework.util.ObjectUtils.isEmpty;
+
+@ContextConfiguration(
+        locations = {
+                "/applicationContext-service.xml",
+                "/applicationContext-cache.xml",
+                "/applicationContext-testdb.xml",
+                "/applicationContext-mockSonos.xml" })
+@DirtiesContext(
+        classMode = DirtiesContext.ClassMode.AFTER_EACH_TEST_METHOD)
+/*
+ * Tests to prove what kind of strings/chars can be used in the genre field.
+ */
+public class SearchServiceSpecialGenreTestCase {
+
+    @ClassRule
+    public static final SpringClassRule classRule = new SpringClassRule() {
+        HomeRule homeRule = new HomeRule();
+
+        @Override
+        public Statement apply(Statement base, Description description) {
+            Statement spring = super.apply(base, description);
+            return homeRule.apply(spring, description);
+        }
+    };
+
+    @Rule
+    public final SpringMethodRule springMethodRule = new SpringMethodRule();
+
+    @Autowired
+    private MediaScannerService mediaScannerService;
+
+    @Autowired
+    private MusicFolderDao musicFolderDao;
+
+    @Autowired
+    private DaoHelper daoHelper;
+
+    @Autowired
+    private SearchService searchService;
+
+    @Autowired
+    private SettingsService settingsService;
+
+    @Rule
+    public TemporaryFolder temporaryFolder = new TemporaryFolder();
+
+    @Autowired
+    ResourceLoader resourceLoader;
+
+    @Before
+    public void setup() throws Exception {
+        populateDatabase();
+    }
+
+    private static boolean dataBasePopulated;
+
+    private static Function<String, String> resolvePath = (childPath) ->{
+        return MusicFolderTestData.resolveBaseMediaPath() + childPath;
+    };
+    
+    private List<MusicFolder> musicFolders;
+    
+    private List<MusicFolder> getTestMusicFolders() {
+        if (isEmpty(musicFolders)) {
+            musicFolders = new ArrayList<>();
+
+            File musicDir = new File(resolvePath.apply("Search/SpecialGenre"));
+            musicFolders.add(new MusicFolder(1, musicDir, "accessible", true, new Date()));
+
+        }
+        return musicFolders;
+    }
+
+    /*
+     * A count to deliberately stagger the execution of populateDatabase.
+     * If populate Database is called from multiple methods almost simultaneously,
+     * it may not execute correctly.
+     */
+    private int count = 1;
+
+    private synchronized void populateDatabase() {
```

Comment:
Rather than doing this, doesn't it make more sense to use unique airsonic home paths for each test method?

Reply:
Also this method seems duplicated across these test classes. Can either move it to a utility class or add some inheritance.

Reply:
Solved by inheritance.

> Rather than doing this, doesn't it make more sense to use unique airsonic home paths for each test method?

If the meaning is not to define MusicFoldsder ​​in each test class, but to define in MusicFolderTestData collectively, modify additionally.

There are some problems when included in the current MusicFolderTestData# getTestMusicFolders, so I think it will be in the form of adding new methods.